### PR TITLE
Ignore casing when looking up Naming Rules/Styles/Symbols

### LINKED ISSUE #40705: Allow editorconfig Naming Rules/Styles/SymbolSpecs to have uppercase letters
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/869465/editorconfig-naming-convention-not-working-c-visua.html)._

---
I am trying to use .editorconfig at solution level to define naming conventions for the two C# projects in it. I have latest Visual Studio as of date, updated just today to 16.4.2. It is giving warnings where not expected. For example, for class names, property names, enum names it says that it should begin with lower case, whereas as per my convention in editorconfig, they are expected in pascal case. Similarly for private fields in class, I have defined field name to begin with underscore ("_") character but it is reporting that "_" is unexpected.

I am sure that it is using my editorconfig file because I see changes in messages if I deliberately change some parts just to see whether there is any impact. However, result is not as per expectation.
Some lines where error is reported but not expected are:
public partial class WebApplication : AnApplication - "The first word, 'WebApplication', must begin with a lower case character."
private readonly string _driverPath; - "Prefix '_' is not expected"
public RemoteWebDriver WebDriver { get; private set; } -  "The first word, 'WebDriver', must begin with a lower case character."

Are there more settings that need to be done? Is there an error in my editorconfig inputs?

Here is the relevant content of my editorconfig file.

```ini
# Naming rules definitions

dotnet_naming_rule.interfacesPascalWithI.symbols = interfaces
dotnet_naming_rule.interfacesPascalWithI.style = IP
dotnet_naming_rule.interfacesPascalWithI.severity = suggestion

dotnet_naming_rule.publicMethodsPascal.symbols = pubMethods
dotnet_naming_rule.publicMethodsPascal.style = P
dotnet_naming_rule.publicMethodsPascal.severity = suggestion

dotnet_naming_rule.defaultMethodsCamel.symbols = methods
dotnet_naming_rule.defaultMethodsCamel.style = c
dotnet_naming_rule.defaultMethodsCamel.severity = suggestion

dotnet_naming_rule.staticFieldsPascal.symbols = staticFields
dotnet_naming_rule.staticFieldsPascal.style = c
dotnet_naming_rule.staticFieldsPascal.severity = warning

dotnet_naming_rule.constFieldsPascal.symbols = constFields
dotnet_naming_rule.constFieldsPascal.style = c
dotnet_naming_rule.constFieldsPascal.severity = warning

dotnet_naming_rule.privateFieldsUnderscore.symbols = prvFields
dotnet_naming_rule.privateFieldsUnderscore.style = _c
dotnet_naming_rule.privateFieldsUnderscore.severity = suggestion

dotnet_naming_rule.defaultFieldsCamel.symbols = fields
dotnet_naming_rule.defaultFieldsCamel.style = c
dotnet_naming_rule.defaultFieldsCamel.severity = suggestion

dotnet_naming_rule.defaultPascal.symbols = default
dotnet_naming_rule.defaultPascal.style = P
dotnet_naming_rule.defaultPascal.severity = suggestion

# Style Definitions
dotnet_naming_style. P.capitalization = pascal_case

dotnet_naming_style. IP.capitalization = pascal_case
dotnet_naming_style. IP.required_prefix = I

dotnet_naming_style.c.capitalization = camel_case

dotnet_naming_style._c.capitalization = camel_case
dotnet_naming_style._c.required_prefix = _

# Symbols Definitions
dotnet_naming_symbols.default.applicable_kinds = *
dotnet_naming_symbols.default.applicable_accessibilities = *

dotnet_naming_symbols.interfaces.applicable_kinds = interface
dotnet_naming_symbols.interfaces.applicable_accessibilities = *

dotnet_naming_symbols.pubMethods.applicable_kinds = method
dotnet_naming_symbols.pubMethods.. applicable_accessibilities = public

dotnet_naming_symbols.methods.applicable_kinds = method
dotnet_naming_symbols.methods.applicable_accessibilities = *

dotnet_naming_symbols.staticFields.applicable_kinds = field
dotnet_naming_symbols.staticFields.applicable_accessibilities = *
dotnet_naming_symbols.staticFields.required_modifiers = static

dotnet_naming_symbols.constFields.applicable_kinds = field
dotnet_naming_symbols.constFields.applicable_accessibilities = *
dotnet_naming_symbols.constFields.required_modifiers = const

dotnet_naming_symbols.prvFields.applicable_kinds = field
dotnet_naming_symbols.prvFields.applicable_accessibilities = private

dotnet_naming_symbols.fields.applicable_kinds = field
dotnet_naming_symbols.fields.applicable_accessibilities = *
```

---
#### Visual Studio Feedback System on 12/30/2019, 01:49 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 12/30/2019, 02:21 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes #40705

### CODE REVIEW

File: src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs [hunk 1]
```diff
@@ -127,7 +127,11 @@ public static NamingStylePreferences ParseDictionary(IReadOnlyDictionary<string,
 
         private static Dictionary<string, string> TrimDictionary(IReadOnlyDictionary<string, string> allRawConventions)
         {
-            var trimmedDictionary = new Dictionary<string, string>(allRawConventions.Count);
+            // At this point, keys have been lowercased. So, for example, to make a naming style
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: @jasonmalinowski It looks like the keys were lowercased long before we got to this feature. Is that expected, or is there a bug deeper in how document options are constructed, or how keys are read initially?

**Comment by dpoeschl** [comment]
FYI @jmarolf @dotnet/roslyn-ide

File: src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: why are keys being lowercased?

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs [hunk 2]
```diff
@@ -79,6 +79,31 @@ public static void TestPascalCaseRule()
...
+            {
+                ["dotnet_naming_rule.methods_and_properties_must_be_pascal_case.severity"] = "warning",
+                ["dotnet_naming_rule.methods_and_properties_must_be_pascal_case.symbols"] = "Method_and_Property_symbols",
+                ["dotnet_naming_rule.methods_and_properties_must_be_pascal_case.style"] = "Pascal_Case_style",
+                ["dotnet_naming_symbols.method_and_property_symbols.applicable_kinds"] = "method,property",
+                ["dotnet_naming_symbols.method_and_property_symbols.applicable_accessibilities"] = "*",
+                ["dotnet_naming_style.pascal_case_style.capitalization"] = "pascal_case"
+            };
+            var result = ParseDictionary(dictionary);
+            Assert.Single(result.NamingRules);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: Assert.Single() returns the item, so you can combine this and the next line, ditto for everything else.

File: src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs (hunk 1 above)
Comment (JoeRobich) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: from https://editorconfig.org 
> Currently all properties and values are case-insensitive. They are lowercased when parsed.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: So I'm not sure where they're being lowercased, but:

https://github.com/dotnet/roslyn/blob/2f57d11ecc36d11c06fc862dcb0de37c96d305d2/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerConfigOptions.cs#L12-L16

That should be the comparer used in any dictionaries being created. It looks like this change dropped that comparer.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: The .editorconfig spec does also specifically [say](https://editorconfig.org/#file-format-details):

> Currently all properties and values are case-insensitive. They are lowercased when parsed.

File: src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs [hunk 3]
```diff
@@ -127,7 +127,11 @@ public static NamingStylePreferences ParseDictionary(IReadOnlyDictionary<string,
...
+            // called "Pascal_Case_style" match up correctly with the key
+            // "dotnet_naming_style.pascal_case_style.capitalization", we have to ignore casing for
+            // lookups in this dictionary.
+            var trimmedDictionary = new Dictionary<string, string>(allRawConventions.Count, StringComparer.OrdinalIgnoreCase);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: This should use the KeyComparer property so we're definitely using the same comparer as before.

File: src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: Ah here you go:

https://github.com/dotnet/roslyn/blob/3892c636a33f7c3f2e27bc596ec6c43168b57854/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs#L167

Comment (dpoeschl) [reply src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs]: > Currently all properties and values are case-insensitive. They are lowercased when parsed.

We kinda depend on the values not being lowercased. 😬

**Pull request merged**